renderdoc/core/resource_manager.h

```diff
@@ -645,18 +648,24 @@ class ResourceManager : public ResourceRecordHandler
   WrappedResourceType GetWrapper(RealResourceType real);
   void RemoveWrapper(RealResourceType real);
 
-  void Prepare_ResourceInitialStateIfNeeded(ResourceId id);
-  void Prepare_ResourceIfActivePostponed(ResourceId id);
+  void Prepare_InitialStateIfPostponed(ResourceId id);
+  void SkipOrPostponeOrPrepare_InitialState(ResourceId id, FrameRefType refType);
 
-  void UpdateLastWriteTime(ResourceId id);
+  void UpdateLastWriteTime(ResourceId id, FrameRefType refType);
   void ResetLastWriteTimes();
   void ResetCaptureStartTime();
+  void UpdateLastNonCompleteWriteTime(ResourceId id, FrameRefType refType);
+  void ResetLastNonCompleteWriteTimes();
 
   bool HasPersistentAge(ResourceId id);
+  bool HasIrrelevantAge(ResourceId id);
   bool IsResourcePostponed(ResourceId id);
-  bool IsResourcePersistent(ResourceId id);
+  bool IsResourceSkipped(ResourceId id);
+  bool ShouldPostpone(ResourceId id);
+  bool ShouldSkip(ResourceId id);
 
   virtual bool IsResourceTrackedForPersistency(const WrappedResourceType &res) { return false; }
+  virtual bool IsResourceTrackedForIrrelevancy(const WrappedResourceType &res) { return false; }
```

Comment:
Currently glClear and glCopyImageSubData specify CompleteWrite during active capture. Do you mean we should change those two cases to specify a different FrameRefType and make sure no GL operations can specify CompleteWrite in the future? Without a command buffer to accumulate references on a resource, how would later draws prevent a fbo attachment from being skipped by a clear?

Reply:
They probably shouldn't be CompleteWrite since they don't necessarily clear or copy the whole image, so it should be PartialWrite unless some checks are done to ensure really the whole image is covered. What I mean though is that as soon as you have a glDraw or any other call after the clear that will mark the resource as non-completewrite and it will update the non-complete write time to latest, preventing the resource from being skipped.

The only reason this system works in vulkan is exactly because references on resources are accumulated inside a command buffer which is essentially "atomic" in terms of its use. If a resource is completely written at the start of a command buffer then any other references inside it don't matter, that command buffer can never access prior versions of the resource. On GL we don't have that batching so every draw that writes to or reads from a resource does not completely write it. In order to have a resource that is only complete-written for many seconds it would have to be only cleared, not ever used in any other way.

Reply:
I understand what you mean now. I was mistaking the skipped resourced I saw in GL to be incorrectly skipped, when in fact those resources were just simply not referenced and will skip serialization anyway.